ls1intum/Artemis
# `General`: Add scheduled notifications ⏰ for released exercises

### Checklist

### Motivation and Context

This change was requested by Stephan to make sure that notifications (and soon also emails) that inform about newly created/started exercises should be sent at the time the exercise is released (not sooner!).


### Description

The notifications (based on the NotificationType EXERCISE_CREATED) are now also sent to students (and not only tutors). (This is also in line with the new notification settings)
I followed the already available scheduling classes and work flows as close as possible.

### Steps for Testing


1. Create any new exercise (as Instructor) and set the release date to ca. 2 minutes from now.
2. Open Artemis as a student (that has the same course).
3. Make sure no notification was sent yet!
4. Wait until the release date and check if a notification arrives.

### Review Progress





- Code Review
- Manual Tests
  
Client Test Build fails to external issue (happens with multiple unrelated PRs) (I did not touch the client)
The codacy issue is based on catching generic exceptions, I decided to stick to the current implementation because the other scheduling services all have this structure.

### Test Coverage
I created a new scheduling service test class. (90+%)
The other files only were changed minimally, mostly inline by replacing the call to the GroupNotificatonService directly with the new Scheduling Service. The new methods inside the already existing scheduling pipelines are also very small and 1to1 like all legacy ones.
